### PR TITLE
DOC:Improve the docstring of DataFrame.iloc()

### PR DESCRIPTION
**Pull request by tuhinmahmud**
Checklist for the pandas documentation sprint (ignore this if you are doing
an unrelated PR):

- [X ] PR title is "DOC: update the <your-function-or-method> docstring"
- [ X] The validation script passes: `scripts/validate_docstrings.py <your-function-or-method>`
- [ X] The PEP8 style check passes: `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ X] The html version looks good: `python doc/make.py --single <your-function-or-method>`
- [X ] It has been proofread on language by another sprint participant

Please include the output of the validation script below between the "```" ticks:

```
(pandas_dev) tuhins-mbp:pandas tuhinm@us.ibm.com$ python scripts/validate_docstrings.py pandas.DataFrame.iloc

################################################################################
###################### Docstring (pandas.DataFrame.iloc)  ######################
################################################################################

Purely integer-location based indexing for selection by position.

``.iloc[]`` is primarily integer position based (from ``0`` to
``length-1`` of the axis), but may also be used with a boolean
array.

Allowed inputs are:

- An integer, e.g. ``5``.
- A list or array of integers, e.g. ``[4, 3, 0]``.
- A slice object with ints, e.g. ``1:7``.
- A boolean array.
- A ``callable`` function with one argument (the calling Series, DataFrame
  or Panel) and that returns valid output for indexing (one of the above)

``.iloc`` will raise ``IndexError`` if a requested indexer is
out-of-bounds, except *slice* indexers which allow out-of-bounds
indexing (this conforms with python/numpy *slice* semantics).

See Also
--------
DataFrame.ix : A primarily label-location based indexer, with integer position fallback.
DataFrame.loc : Fast integer location scalar accessor.

Examples
--------
>>> import pandas as pd
>>> mydict = [{'a': 1, 'b': 2, 'c': 3, 'd': 4},
...           {'a': 100,  'b': 200, 'c': 300, 'd': 400},
...           {'a': 1000,  'b': 2000,  'c': 3000,  'd': 4000 }]
>>> df = pd.DataFrame(mydict)
>>> print(df.head())
      a     b     c     d
0     1     2     3     4
1   100   200   300   400
2  1000  2000  3000  4000
>>> print(df.iloc[0])
a    1
b    2
c    3
d    4
Name: 0, dtype: int64
>>> print(df.iloc[0:2])
     a    b    c    d
0        1    2    3    4
1  100  200  300  400

ref:`Selection by Position <indexing.integer>`

################################################################################
################################## Validation ##################################
################################################################################

Errors found:
	No returns section found



```

If the validation script still gives errors, but you think there is a good reason
to deviate in this case (and there are certainly such cases), please state this
explicitly.

Error Returned because Class do not need return

### CODE REVIEW

File: pandas/core/indexing.py [hunk 1]
```diff
@@ -1648,8 +1649,35 @@ class _iLocIndexer(_LocationIndexer):
...
+    b    2
+    c    3
+    d    4
+    Name: 0, dtype: int64
+    >>> print(df.iloc[0:2])
+         a    b    c    d
+    0        1    2    3    4
+    1  100  200  300  400
+
+    ref:`Selection by Position <indexing.integer>`
```
Comment (TomAugspurger) [inline pandas/core/indexing.py]: Could you use the "extended summary" section for this, right below the opening summary? I think the prose docs are the most important for indexing, and I don't want them to get lost.

File: pandas/core/indexing.py [hunk 2]
```diff
@@ -1648,8 +1649,35 @@ class _iLocIndexer(_LocationIndexer):
     out-of-bounds, except *slice* indexers which allow out-of-bounds
     indexing (this conforms with python/numpy *slice* semantics).
 
-    See more at :ref:`Selection by Position <indexing.integer>`
+    See Also
+    --------
+    DataFrame.ix : A primarily label-location based indexer
```
Comment (TomAugspurger) [inline pandas/core/indexing.py]: ix is deprecated, you can remove.

File: pandas/core/indexing.py [hunk 3]
```diff
@@ -1648,8 +1649,35 @@ class _iLocIndexer(_LocationIndexer):
...
+    DataFrame.loc : Fast integer location scalar accessor.
```
Comment (TomAugspurger) [inline pandas/core/indexing.py]: I think you may want the " A primarily label-location based indexer" for the .loc description.

You may be thinking of `DataFrame.iat` for this dsecription.

File: pandas/core/indexing.py [hunk 4]
```diff
@@ -1648,8 +1649,35 @@ class _iLocIndexer(_LocationIndexer):
...
 
+    Examples
+    --------
+    >>> import pandas as pd
+    >>> mydict = [{'a': 1, 'b': 2, 'c': 3, 'd': 4},
+    ...           {'a': 100,  'b': 200, 'c': 300, 'd': 400},
```
Comment (TomAugspurger) [inline pandas/core/indexing.py]: PEP8: single spaces.

File: pandas/core/indexing.py [hunk 5]
```diff
@@ -1648,8 +1649,35 @@ class _iLocIndexer(_LocationIndexer):
...
+    2  1000  2000  3000  4000
+    >>> print(df.iloc[0])
+    a    1
+    b    2
+    c    3
+    d    4
+    Name: 0, dtype: int64
+    >>> print(df.iloc[0:2])
+         a    b    c    d
+    0        1    2    3    4
```
Comment (TomAugspurger) [inline pandas/core/indexing.py]: formatting seems a bit off. Can you double check this?

**Review by TomAugspurger — CHANGES_REQUESTED** [review]

File: pandas/core/indexing.py [hunk 6]
```diff
@@ -1648,8 +1649,35 @@ class _iLocIndexer(_LocationIndexer):
...
+    0     1     2     3     4
+    1   100   200   300   400
+    2  1000  2000  3000  4000
+    >>> print(df.iloc[0])
+    a    1
+    b    2
+    c    3
+    d    4
+    Name: 0, dtype: int64
+    >>> print(df.iloc[0:2])
```
Comment (jreback) [inline pandas/core/indexing.py]: don't need the prints here and above

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexing.py (hunk 4 above)
Comment (tuhinmahmud) [reply pandas/core/indexing.py]: updated the double spaces with single

File: pandas/core/indexing.py (hunk 5 above)
Comment (tuhinmahmud) [reply pandas/core/indexing.py]: reran the command and updated.

File: pandas/core/indexing.py (hunk 6 above)
Comment (tuhinmahmud) [reply pandas/core/indexing.py]: took off the prints

File: pandas/core/indexing.py (hunk 3 above)
Comment (tuhinmahmud) [reply pandas/core/indexing.py]: Remove ix and update loc description:

-    DataFrame.ix : A primarily label-location based indexer
-    DataFrame.loc : Fast integer location scalar accessor.
+    DataFrame.iat : Fast integer location scalar accessor.
+    DataFrame.loc : Purely label-location based indexer for selection by label.

Comment (tuhinmahmud) [reply pandas/core/indexing.py]: missed your comment about .loc .. but instead got the description from https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.loc.html

File: pandas/core/indexing.py (hunk 1 above)
Comment (tuhinmahmud) [reply pandas/core/indexing.py]: moved it to "extended summary"

**Review by tuhinmahmud — COMMENTED** [review]
ok

File: pandas/core/indexing.py (hunk 2 above)
Comment (tuhinmahmud) [reply pandas/core/indexing.py]: done

File: pandas/core/indexing.py [hunk 7]
```diff
@@ -1642,14 +1643,41 @@ class _iLocIndexer(_LocationIndexer):
...
     out-of-bounds, except *slice* indexers which allow out-of-bounds
     indexing (this conforms with python/numpy *slice* semantics).
 
-    See more at :ref:`Selection by Position <indexing.integer>`
+    ref:`Selection by Position <indexing.integer>`.
+
+    See Also
+    --------
+    DataFrame.iat : Fast integer location scalar accessor.
+    DataFrame.loc : Purely label-location based indexer for selection by label.
```
Comment (jreback) [inline pandas/core/indexing.py]: add Series.iloc

File: pandas/core/indexing.py [hunk 8]
```diff
@@ -1642,14 +1643,41 @@ class _iLocIndexer(_LocationIndexer):
...
 
+    Examples
+    --------
+    >>> import pandas as pd
```
Comment (jreback) [inline pandas/core/indexing.py]: don't need the pandas import

File: pandas/core/indexing.py [hunk 9]
```diff
@@ -1642,14 +1643,41 @@ class _iLocIndexer(_LocationIndexer):
...
+    >>> mydict = [{'a': 1, 'b': 2, 'c': 3, 'd': 4},
+    ... {'a': 100, 'b': 200, 'c': 300, 'd': 400},
```
Comment (jreback) [inline pandas/core/indexing.py]: need to indent here

File: pandas/core/indexing.py [hunk 10]
```diff
@@ -1642,14 +1643,41 @@ class _iLocIndexer(_LocationIndexer):
...
+    ... {'a': 1000, 'b': 2000, 'c': 3000, 'd': 4000 }]
+    >>> df = pd.DataFrame(mydict)
+    >>> df
+          a     b     c     d
+    0     1     2     3     4
+    1   100   200   300   400
+    2  1000  2000  3000  4000
+    >>> df.iloc[0]
```
Comment (jreback) [inline pandas/core/indexing.py]: blank lines between cases. 

show additional examples, including selecting with .iloc[0] vs .iloc[[0]], and use a multi-axis selction .iloc[0, 0] and lists for the last, e.g. .iloc[[0, 1], [0, 1]]

Comment (jreback) [reply pandas/core/indexing.py]: also add a sentence for each case explaining it.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexing.py (hunk 10 above)
Comment (tuhinmahmud) [reply pandas/core/indexing.py]: added sentences and put different types of example of .iloc in paragraph

File: pandas/core/indexing.py (hunk 9 above)
Comment (tuhinmahmud) [reply pandas/core/indexing.py]: undated the indentation

File: pandas/core/indexing.py (hunk 8 above)
Comment (tuhinmahmud) [reply pandas/core/indexing.py]: removed pandas import

File: pandas/core/indexing.py (hunk 7 above)
Comment (tuhinmahmud) [reply pandas/core/indexing.py]: added Series.iloc

**Comment by TomAugspurger** [comment]
@tuhinmahmud pushed an update to your examples to make the ordering a bit more logical. We now show each of the valid indexer values (scalar, sequence, slice, mask, callable) for just indexing the rows. Then we show the same for rows and columns.

**Comment by TomAugspurger** [comment]
fireshot capture 001 - pandas dataframe iloc pandas 0 23 0_ - file____users_taugspurger_sandbox_

File: pandas/core/indexing.py (hunk 7 above)
Comment (jorisvandenbossche) [reply pandas/core/indexing.py]: Can you look at other PRs for those accessors to have a consistent way to reference them? See eg https://github.com/pandas-dev/pandas/pull/20229/files, They use:

```
    DateFrame.at : Access a single value for a row/column label pair
    DateFrame.iloc : Access group of rows and columns by integer position(s)
    Series.loc : Access group of values using labels
```

Comment (TomAugspurger) [reply pandas/core/indexing.py]: @tuhinmahmud make sure to pull my changes before looking into this.

**Review by tuhinmahmud — COMMENTED** [review]
made requested changes
i) Remove ix and update loc description
ii) removed pandas import
iii) single space issue

**Review by tuhinmahmud — COMMENTED** [review]
Looks good.. Do you know what is the next step to get it commited?

**Comment by TomAugspurger** [comment]
Merged in master to restart the CI. They were complaining about a non-ASCII character, but that may have been on master when you made the PR, and not in your actual branch.

Ping if you notice that the tests all pass before we do.

**Comment by tuhinmahmud** [comment]
@TomAugspurger  Hi .. I am kind of lost as to what I can do to make it work .. I see one of the test continuous-integration/travi-ci/pr failing.. not sure what to make of it 

image

**Comment by TomAugspurger** [comment]
Sometimes travis jobs take much longer than others, and exceed Travis' time limit. I restarted that one.

**Comment by mroeschke** [comment]
Looks like everything was addressed. Will merge on green.

**Pull request merged**